internal/core/application/blockchain_listener.go

```diff
@@ -67,98 +75,172 @@ func newBlockchainListener(
-				err.Error(),
-			)
-			break
-		}
-	case crawler.TransactionConfirmed:
-		e := event.(crawler.TransactionEvent)
-		ctx := context.Background()
-		if _, err := b.dbManager.RunTransaction(
-			ctx,
-			!readOnlyTx,
-			func(ctx context.Context) (interface{}, error) {
-				return nil, b.updateTrade(ctx, e)
-			},
-		); err != nil {
-			log.Warnf(
-				"trying to update trade completion status %s\n",
-				err.Error(),
-			)
-			break
+	observable := &crawler.TransactionObservable{TxID: txid}
+
+	if !b.started {
+		b.pendingObservables = append(b.pendingObservables, observable)
+		return
+	}
+	b.crawlerSvc.AddObservable(observable)
+}
+
+func (b *blockchainListener) StopObserveAddress(addr string) {
+	b.crawlerSvc.RemoveObservable(&crawler.AddressObservable{
+		Address: addr,
+	})
+}
+
+func (b *blockchainListener) StopObserveTx(txid string) {
+	b.crawlerSvc.RemoveObservable(&crawler.TransactionObservable{TxID: txid})
+}
+
+func (b *blockchainListener) listentToEventChannel() {
+	for {
+		select {
+		case event := <-b.crawlerSvc.GetEventChannel():
+			switch event.Type() {
+			case crawler.CloseSignal:
+				log.Debug("CloseEvent detected")
+				log.Debug("stop listening on event channel")
+				return
+			case crawler.FeeAccountDeposit:
+				ctx, err := b.handleUnspents(event)
+				if err != nil {
+					break
+				}
+				if _, err := b.dbManager.RunTransaction(
+					ctx,
+					readOnlyTx,
+					func(ctx context.Context) (interface{}, error) {
+						return nil, b.checkFeeAccountBalance(ctx)
+					},
+				); err != nil {
+					log.Warnf(
+						"trying to check balance for fee account: %s\n",
+						err.Error(),
+					)
+					break
+				}
+
+			case crawler.MarketAccountDeposit:
+				ctx, err := b.handleUnspents(event)
+				if err != nil {
+					break
+				}
+				e := event.(crawler.AddressEvent)
+				if _, err := b.dbManager.RunTransaction(
+					ctx,
+					!readOnlyTx,
+					func(ctx context.Context) (interface{}, error) {
+						return nil, b.checkMarketAccountFundings(ctx, e.AccountIndex)
+					},
+				); err != nil {
+					log.Warnf(
+						"trying to check fundings for market account %d: %s\n",
+						e.AccountIndex,
+						err.Error(),
+					)
+					break
+				}
+			case crawler.TransactionConfirmed:
+				e := event.(crawler.TransactionEvent)
+				ctx := context.Background()
+				if _, err := b.dbManager.RunTransaction(
+					ctx,
+					!readOnlyTx,
+					func(ctx context.Context) (interface{}, error) {
+						return nil, b.updateTrade(ctx, e)
+					},
+				); err != nil {
+					log.Warnf(
+						"trying to update trade completion status %s\n",
+						err.Error(),
+					)
+					break
+				}
+				// stop watching for a tx after it's confirmed
+				b.StopObserveTx(e.TxID)
+			}
+		default:
 		}
 	}
 }
 
+func (b *blockchainListener) startPendingObservables() {
+	if len(b.pendingObservables) <= 0 {
+		return
+	}
+
+	for _, observable := range b.pendingObservables {
+		b.crawlerSvc.AddObservable(observable)
+		time.Sleep(200 * time.Millisecond)
```

Comment:
Do you think is 200ms is too much or enought? With Explorer service for sure too much, with ELements too little?

Reply:
do we have exact data on limitation from Explorer/Elements?

Reply:
this kind of time.Sleep prevents the observations to be almost concurrent basically... consider that usually the daemon adds more than one address to the watchlist... so these additions are "mitigated" by the sleeping time.

Reply:
@sekulicd  on the Blockstream.info instance we don't know what they do and they can change unilaterally at they discretion, ban our IPs and so on :P So do not relay on that as info

On the new "Elements as poor man explorer" instead It depends on how you configure the Elements node, but we do not need to expect an optimized node for high availability & high load enviroment.

Those the available conf options

``` 
-rpcthreads=<n>

Set the number of threads to service RPC calls (default: 4) 
       
-rpcworkqueue=<n>

Sets the depth of the work queue to service RPC calls, the default is 16.
```

Reply:
we defenitely know to not "pump" your elements node too much or it will crash :D

https://github.com/bitpay/bitcore-node/issues/463#issuecomment-418575795

Reply:
check the [comment right below](https://github.com/bitpay/bitcore-node/issues/463#issuecomment-418592081)... the crash was due to another reason 😅

Reply:
Ok cause I found via wasabi https://github.com/zkSNACKs/WalletWasabi/pull/3056